guidance/_parser.py

```diff
@@ -1,666 +1,302 @@
-            rs = ""
-            for i, v in enumerate(self.values):
-                if self.pos == i:
-                    rs += "•"
-                rs += v.name + " "
-            if self.pos == len(self.values):
-                rs += "•"
-        elif isinstance(self.node, Select):
-            s = f"{self.node.name:20} -> "
-            rs = ""
-            if self.pos == 0:
-                rs += "•"
-            rs += self.values[0].name
-            if self.pos == 1:
-                rs += "•"
-        else:
-            assert False
-        return s + f"{rs:40} ({self.start}) {'nullable' if self.node.nullable else ''}"
 
+class TokenParserException(Exception):
+    pass
 
-class Parser:
-    """An abstract base class for guidance parsers."""
 
-    pass
+class InvalidTokenException(TokenParserException):
+    def __init__(self, token: int, valid_tokens: list[int], prompt_tokens: list[int]):
+        self.token = token
+        self.valid_tokens = valid_tokens
+        self.prompt_tokens = prompt_tokens
+        super().__init__(
+            f"Invalid token {token}, expected one of {valid_tokens} after {prompt_tokens}"
+        )
 
 
-class EarleyCommitParser(Parser):
-    def __init__(self, grammar):
+class TokenParser:
+
+    def __init__(
+        self,
+        grammar: Union[GrammarFunction, str],
+        tokenizer: Tokenizer,
+        prompt: bytes = b"",
+        ensure_bos_token: bool = True,
+    ):
+        if isinstance(grammar, GrammarFunction):
+            # we can't have a terminal as the root
+            if isinstance(grammar, Terminal):
+                grammar = Join([grammar])
+            serialized_grammar = json.dumps(grammar.ll_serialize())
+        else:
+            serialized_grammar = grammar
 
-        # we can't have a terminal as the root
-        if isinstance(grammar, Terminal):
-            grammar = Join([grammar])
+        self.tokenizer = tokenizer
+        self.ll_tokenizer = llguidance.LLTokenizer(
+            llguidance.TokenizerWrapper(tokenizer)
+        )
+        self.ll_interpreter = llguidance.LLInterpreter(
+            self.ll_tokenizer,
+            serialized_grammar,
+            log_level=int(os.environ.get("LLGUIDANCE_LOG_LEVEL", "1")),
+        )
+        self._generator = self._parse(prompt, ensure_bos_token)
+        self._done = False
+
+    def is_accepting(self) -> bool:
+        return self.ll_interpreter.is_accepting()
+
+    def done(self) -> bool:
+        return self._done
+
+    def advance(
+        self, token: Optional[int]
+    ) -> Tuple[Optional[GenData], EngineCallResponse]:
+        try:
+            return self._generator.send(token)
+        except StopIteration as e:
+            self._done = True
+            return None, e.value
+
+    def _process_prompt(self, prompt: bytes, ensure_bos_token: bool) -> list[int]:
+        prompt_tokens = self.ll_interpreter.process_prompt(
+            self.tokenizer.encode(prompt)
```

Comment:
Is there a reason to use `self.tokenizer.encode(prompt)` instead of `self.ll_tokenizer.tokenize_bytes(prompt)`? Would one be faster, and is the tokenization the same?

Reply:
Ah, good catch -- I don't think I knew that method existed ;)
I'd guess that the latter would be more "idiomatic" for llguidance.
@mmoskal any comment?

Reply:
The tokenize_bytes will probably not add the BOS? Please check!